### DISCUSSION TITLE
Improve struct constructor to make it work like class constructor

### DISCUSSION THREAD

**Discussion by KorneiDontsov** [dotnet/csharplang#3420] [Language Ideas]
When I use `struct` and wrote parameterized constructor, I must initialize every field with a value, even if it is going to be `default`.

```c#
struct MyStruct {
    public readonly Object required;

    public Object? optional;

    public MyStruct(Object required) {
        this.requred = required;
        optional = null;
    }
}
```

But in `class` I doesn't have to do it. Field is `default`/`null` by default.

It's annoying that in `struct` I have to wrote even more code than in `class` in order to do exactly same thing. I propose to make the compiler do the work even if implementation of how `struct` constructor works is different from how `class` constructor works.

## Examples

```c#
struct MyStruct {
    public readonly Object required;

    public Object? optional;

    public MyStruct(Object required) {
        this.requred = required;
        // Compiler understands that 'optional' is not initialized
        // and initializes it here with null.
    }
}
```

```c#
struct MyStruct {
    public readonly Object required;

    public Object? optional;

    public MyStruct(Object required) {
        this.requred = required;
        // Compiler understands that 'optional' is going to be used without being initialized 
        // and initializes it here with null.
        Console.WriteLine(optional);
        // or
        Console.WriteLine(this);
    }
}
```

**Comment by SingleAccretion** [comment]
Note that there is a workaround for this limitation:
````CSharp
public struct Struct
{
    public int Int;    
    public double Double;
    public string String;
    
    public Struct(int integer)
    {
        this = default;
        Int = integer;
    }        
}
````
It is a bit ugly, but it does save on typing.

**Comment by svick** [comment]
Another workaround, arguably one that is less ugly, is to [explicitly delegate to `this()`](https://sharplab.io/#v2:C4LgbgpgTglgZgTwD4GdhQK4GNgAIDKAsAFADeJuluAAgMy4wB2eAkswNxW4VV24AmAewwAjADYRcAEWHiI7HpT7UAjAAYC6JgHMFxLopr18ACiZ5zEbdACUuELmAALGChM3D5fVyps8AXgZmK2g9LgBfHyoScKA):

```c#
struct S
{
    public int Int;    
    public double Double;
    public string String;
    
    public S(int integer) : this()
    {
        Int = integer;
    }        
}
```
